src/toil_scripts/gatk_germline/germline.py

```diff
@@ -1,304 +1,500 @@
+        except AttributeError:
+            setattr(config, name, None)
+        finally:
+            if getattr(config, name, None) is None and name not in unrequired_references:
+                raise ValueError("Necessary parameter is missing:\n{}".format(name))
+    return job.addFollowOnJobFn(reference_preprocessing, config).rv()
+
+
+def reference_preprocessing(job, config):
     """
     Creates a genome.fa.fai and genome.dict file for reference genome.fa
     if one is not already present in the config dictionary
 
-    :param config dict: pipeline configuration options and shared files
-                        requires a key called 'genome.fa'
-    :return dict: updated config dictionary with reference index files
+    :param JobFunctionWrapping Job: passed by Toil
+    :param Namespace config: Pipeline configuration options and shared files.
+                             Requires genome_fasta attribute
+    :return: Updated config with reference index files
+    :rtype: Namespace
     """
     job.fileStore.logToMaster('Preparing Reference Files')
-    genome_id = config['genome.fa']
-    if 'genome.fa.fai' not in config or config['genome.fa.fai'] is None:
-        config['genome.fa.fai'] = job.addChildJobFn(run_samtools_faidx, genome_id, mock=mock).rv()
-    if 'genome.dict' not in config or config['genome.dict'] is None:
-        config['genome.dict'] = job.addChildJobFn(run_picard_create_sequence_dictionary,
-                                                  genome_id, mock=mock).rv()
+    genome_id = config.genome_fasta
+    if not hasattr(config, 'genome_fai') or config.genome_fai is None:
+        config.genome_fai = job.addChildJobFn(run_samtools_faidx,
+                                              genome_id,
+                                              cores=config.cores).rv()
+    if not hasattr(config, 'genome_dict') or config.genome_dict is None:
+        config.genome_dict = job.addChildJobFn(run_picard_create_sequence_dictionary,
+                                               genome_id,
+                                               cores=config.cores).rv()
     return config
 
 
-def gatk_haplotype_caller(job, bam_id, bai_id, config, annotations=None):
+def prepare_bam(job, uuid, url, config, rg_line=None):
+    """
+    Prepares BAM file for GATK Germline pipeline.
+
+    0: Align FASTQ or Download BAM
+    1: Remove secondary alignments
+    2: Sort
+    3: Index
+
+    :param JobFunctionWrappingJob job: Toil Job instance
+    :param str uuid: Unique identifier for the sample
+    :param str url: URL to BAM file
+    :param Namespace config: Configuration options for pipeline
+    :param str rg_line: RG line for BWA alignment. Default is None.
+    :return: BAM and BAI FileStoreIDs
+    :rtype: tuple
+    """
+    # Determine extension on input file
+    base, ext1 = os.path.splitext(url)
+    _, ext2 = os.path.splitext(base)
+
+    # Produce or download BAM file
```

Comment:
Also, minor nit, but ideally you'd support .fq.gz/.fastq.gz/etc and variants with different case (e.g., .FQ)